extractKeyValuePairs function

arthurpassos opened the pull request:
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add `extractKeyValuePairs` function to extract key value pairs from strings. Input strings might contain noise (i.e log files / do not need to be 100% formatted in key-value-pair format), the algorithm will look for key value pairs matching the arguments passed to the function. As of now, function accepts the following arguments: `data_column` (mandatory), `key_value_pair_delimiter` (defaults to `:`), `pair_delimiters` (defaults to `\space \, \;`) and `quoting_character` (defaults to double quotes).

The `extractKeyValuePairsWithEscaping` has also been added and it has escape sequences support built-in. Supported escape sequences are the same as the ones supported by `DB::parseComplexEscapeSequence` as it is used to parse them.

Examples:

**Simple case**
``` sql
arthur :) select extractKeyValuePairs('name:neymar, age:31 team:psg,nationality:brazil') as kv

SELECT extractKeyValuePairs('name:neymar, age:31 team:psg,nationality:brazil') as kv

Query id: f9e0ca6f-3178-4ee2-aa2c-a5517abb9cee

┌─kv──────────────────────────────────────────────────────────────────────┐
│ {'name':'neymar','age':'31','team':'psg','nationality':'brazil'}        │
└─────────────────────────────────────────────────────────────────────────┘
```

**Single quote as quoting character**
``` sql
arthur :) select extractKeyValuePairs('name:\'neymar\';\'age\':31;team:psg;nationality:brazil,last_key:last_value', ':', ';,', '\'') as kv

SELECT extractKeyValuePairs('name:\'neymar\';\'age\':31;team:psg;nationality:brazil,last_key:last_value', ':', ';,', '\'') as kv

Query id: 0e22bf6b-9844-414a-99dc-32bf647abd5e

┌─kv───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ {'name':'neymar','age':'31','team':'psg','nationality':'brazil','last_key':'last_value'}                                 │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

**Escape sequences without escape sequences support**
``` sql
arthur :) select extractKeyValuePairs('age:a\\x0A\\n\\0') as kv

SELECT extractKeyValuePairs('age:a\\x0A\\n\\0') AS kv

Query id: e9fd26ee-b41f-4a11-b17f-25af6fd5d356

┌─kv────────────────────┐
│ {'age':'a\\x0A\\n\\0'} │
└───────────────────────┘
```
**Escape sequences with escape sequence support turned on**
``` sql
arthur :) select extractKeyValuePairsWithEscaping('age:a\\x0A\\n\\0') as kv

SELECT extractKeyValuePairsWithEscaping('age:a\\x0A\\n\\0') AS kv

Query id: 44c114f0-5658-4c75-ab87-4574de3a1645

┌─kv───────────────┐
│ {'age':'a\n\n\0'} │
└──────────────────┘
```


**Implementation design**:

Input string is processed by a state machine, see `CHKeyValuePairExtractor` and `StateHandlerImpl`.

If escaping support is off, there is no temporary copies of possible keys and values. `std::string_view` is used to store references and later on they are inserted directly into the output column. If escaping support is on, possible keys and values are copied directly into the output columns. If those turn out to be invalid keys/ values, they are discarded without de-allocating the memory used by them. Check `NoEscapingStateHandler::StringWriter` and `InlineEscapingStateHandler::StringWriter` for reference.

SSE42 and SSE2 are used for string search when available on the system. For that, runtime versions of `find_first_symbols` had to be added to `base/find_symbols` as the needles/ symbols to be searched for are not known at compile time.

**Functional (user perspective) differences between previous implementation (last reviewed) and current implementation**

1. Order of input parameters have changed.
2. No more `special_value_characters_allowlist`.
3. As long as a character is not a control character it's considered valid, as opposed to previous implementation where only alphanum was considered non-special and everything else had to be escaped. In pratice, that means !@#$%¨&*( and other "special" characters will be considered valid part of keys and values unless specified in control characters.
4. Default escape sequences are now transformed into their corresponding values (e.g \n -> 0xA).
5. Non default escape sequences are now returned as it is (e.g \\$ -> \\$).
6. There can be multiple pair delimiters as opposed to only one. By default \space is one of them, which means `name :arthur` will not output anything because it has found a pair delimiter where it is not supposed to. If space is removed from delimiters, the output key will include the trailing space.
7. Escape character is no longer configurable, it's hardcoded to `\`.
 

Closes https://github.com/ClickHouse/ClickHouse/issues/41009



> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

arthurpassos committed:
tmp commit

arthurpassos committed:
add value allowlist

arthurpassos committed:
Move everything to DB namespace

arthurpassos committed:
progress

arthurpassos committed:
progress

arthurpassos committed:
progress

arthurpassos committed:
add tests & incorporate ch replace function

arthurpassos committed:
small change

vdimir commented on src/Functions/keyvaluepair/src/KeyValuePairExtractor.h:
```diff
@@ -0,0 +1,30 @@
+namespace DB
+{
+/*
+ * Extracts key value pairs from strings. Strings do not need to be key-value pair only,
```
Before finishing the code, we can discuss examples of how it would work from the user side.

vdimir replied:
Initially, I overlooked `gtest_key_value_pair_extractor.cpp` and wanted to see any examples to understand what is `reserved_char` for example. But I've checked these examples, and they look sane.

For now, the question is, what are the escaping rules for quoted string? E.g., are the escaping of control chars applied ( `"\n"` ) ? And how are quotes handled (with slash `"with \"quoted\" inside"`, or csv-style double quotes `"with ""quotes"" inside"`)?

Also, are there any plans to support the automatic detection of separators? This function tries to do best-effort parsing and be flexible, so it may work like that. So all `parserKV('aa: one,  bb: two,'), parserKV('aa=one bb=two'), parserKV('aa => one; bb => two')` would return `{"aa: "one", "bb": "two"}`. For that, we can have some list of possible kv_separators (`':', '=', '=>', ...`) and item delimiters (`',', ' ', ';', '|', ...`) and check what combination is appeared more frequently. But still trying to figure out if it would be helpful or if it would be challenging to implement and use then.

arthurpassos replied:
1. **For now, the question is, what are the escaping rules for quoted string?** - _I am not sure I understand the question, but escaping does not happen on quoted strings. It'll read everything in until it gets another unescaped quote._
2. **Are the escaping of control chars applied ( "\n" ) ?** - _It'll read everything in, which means the output will contain the back-slash._
3.  **And how are quotes handled (with slash "with \"quoted\" inside", or csv-style double quotes "with ""quotes"" inside")?** - _Quotes, just like any other special character, need to be escaped. The escaping character is configurable, but it defaults to back-slash (\\)._
4. **Also, are there any plans to support the automatic detection of separators?** - _It is not on my plans and I think we should keep the first version simpler. I can totally implement it in the future or as a follow up of this PR if we decide it's actually useful._

I still haven't had the time to look into parameter cross validation (e.g user setting back-slash as both key_value_delimiter and escape character).

vdimir replied:
> It'll read everything in until it gets another unescaped quote

So, having quote `"`" in quoted text is not possible?

arthurpassos replied:
It is possible, but it needs to be escaped. Assuming the escaping character is '\',  an input like: "random_key"="aaa\\"bbb" will output: "random_key"="aaa"bbb". The quote in the middle being a value character, not a delimiter

arthurpassos committed:
rename methods

arthurpassos committed:
reduce state management in lazyescapingkvpextractor by keeping part of it in state handlers

arthurpassos committed:
Make KeyValuePairExtractor response generic and introduce no-op escaping processor

arthurpassos committed:
first test, big file

arthurpassos committed:
rename to exrtactKeyValue and improve code

arthurpassos committed:
finish renaming to extractkeyvaluepairs

arthurpassos committed:
fix tests

arthurpassos committed:
hopefully I reformatted everything

arthurpassos committed:
formatting challenge

arthurpassos committed:
formatting

arthurpassos committed:
fix typo and remove double whitespaces from test strings just to make formatter happy

arthurpassos committed:
add tests

arthurpassos commented:
@vdimir I think it's ready for a 1st review.

Pending things:
1. Documentation
2. Parameter validation (e.g duplicated control characters)
3. Possibly improve code of `DB::ExtractKeyValuePairs`
4. Possibly more tests
5. Maybe some benchmarks

arthurpassos committed:
style

arthurpassos commented:
@filimonov

arthurpassos commented:
@vdimir Would you be so kind and you enable CI?

arthurpassos committed:
add en lorem ipsum docs to get a build in CI

arthurpassos committed:
yet another loremipsum

arthurpassos committed:
throw exception on no arguments

arthurpassos committed:
improve docs

arthurpassos committed:
remove double whitespaces to make style happy

arthurpassos committed:
remove redundant base class constructor call

vdimir commented on src/Functions/keyvaluepair/src/impl/state/State.h:
```diff
@@ -0,0 +1,34 @@
+
+enum State
+{
+    WAITING_KEY,
+    READING_KEY,
+    READING_ENCLOSED_KEY,
+    READING_KV_DELIMITER,
+    WAITING_VALUE,
+    READING_VALUE,
+    READING_ENCLOSED_VALUE,
+    READING_EMPTY_VALUE,
+    FLUSH_PAIR,
+    END
```
Let's add doc comment with explanation for each state. E.g. what is `enclosed_key/value`, `flush`, the difference between waiting and reading?

vdimir commented on src/Functions/keyvaluepair/src/impl/state/State.h:
```diff
@@ -0,0 +1,34 @@
+
+struct NextState
+{
+    std::size_t pos;
```
What is pos?

arthurpassos replied:
position in string, will rename it

vdimir commented on src/Functions/keyvaluepair/src/impl/state/State.h:
```diff
@@ -0,0 +1,34 @@
+struct NextStateWithElement
+{
+    NextState state;
+    std::string_view element;
```
What is element?

arthurpassos replied:
This struct is not even used anymore, forgot to delete it. In any case, an element is either a key or a value.

vdimir commented on src/Functions/keyvaluepair/src/KeyValuePairExtractor.h:
```diff
@@ -0,0 +1,32 @@
+ * Both key and values accepts underscores as well. Special characters must be escaped.
+ * Control characters (key_value_pair_separator, item_delimiter, escape_character and enclosing_character) are customizable
+ * */
+template <typename R = std::unordered_map<std::string, std::string>>
```
Do we use R different from `std::unordered_map<std::string, std::string>` or `std::unordered_map<std::string_view, std::string_view>`. if not, can we make it without this parametrization?

arthurpassos replied:
Sort of. When I initially wrote this code, I assumed all extraction & escaping would be done on the lib side and later on I would just insert it on a CH column. Then I wrote unit tests for the lib and the unit tests expect the pairs to be properly escaped.

Since using `DB::ReplaceString` seems to be more optimal for escaping, I had to find a way to preserve the library unit tests & allow `DB::ReplaceString` to be used for escaping. Therefore, the output is parameterized based on the EscapingProcessor, which can be either NoOp (used by CH itself) or SimpleEscaping (used by the library unit tests).

One thing I would like to avoid is coupling the library to ClickHouse (i.e adding ClickHouse dependencies inside the key-value-pair extraction code).

From the top of my head, the options we have are:

1. **Keep it the way it is**. _I do understand it's not very readable and might be confusing/ over engineering, but it doesn't couple library with CH and unit tests can still expect a properly escaped output_.
2. **Remove parametrization, return `std::unordered_map<std::string_view, std::string_view>` by default and create a an Adapter/Decorator interface that will implement simple escaping and return `std::unordered_map<std::string, std::string>` to be used on unit tests**. _With this approach, both CH client code & library code will be simpler. No need for fancy templates and the unit test/ escaping problem will be solved by an adapter/decorator that'll only be used by unit tests._
3. **Remove parametrization and change expectations for key value pair extraction library. It'll no longer be expected to escape characters** _With this approach, both CH client code & library code will be simpler, but the library is not fully function & isolated._

Let me know if you agree with one of these or have a better suggestion

vdimir replied:
So, we use `std::string_view` as much as possible to avoid copies (and in ClickHouse function implementation), but for unit-test it should be escaped that's why copying to `string` is required?

arthurpassos replied:
Yes, it's only copied on unit tests because it needs to be escaped. When the function is invoked in CH, the output of the library is `std::unordered_map<std::string_view, std::string_view>` and then CH escapes/ copies it using `DB::ReplaceString`

vdimir replied:
Let's just add it into the comment

vdimir commented on src/Functions/keyvaluepair/src/KeyValuePairExtractorBuilder.h:
```diff
@@ -0,0 +1,78 @@
+{
+
+template <typename Response = std::unordered_map<std::string, std::string>>
+class KeyValuePairExtractorBuilder
```
Consider making a simple struct with public fields and setting them explicitly in caller without `with...` methods (if the amount of code won't much greater)

arthurpassos replied:
I honestly feel like a Builder presents a much nicer & fluent interface. Plus, it's a very well known & established pattern. Do you think we could keep this as the way it is or it's a no go for you?

arthurpassos replied:
I honestly feel like a Builder presents a much nicer & fluent interface. Plus, it's a very well known & established pattern. Do you think we could keep this as the way it is or it's a no go for you?

vdimir replied:
Ok, let's keep it

vdimir commented on src/Functions/keyvaluepair/src/impl/state/StateHandler.h:
```diff
@@ -0,0 +1,25 @@
+namespace DB
+{
+
+struct StateHandler
```
What is the purpose of `StateHandler`? From the first look, it has only settings (`escape_character`, `enclosing_character`) and `get()` method without reading anything, and it's not clear what the responsibility is. Should children introduce methods that would affect the state?

arthurpassos replied:
Hmm... I guess it's easier to explain starting with the derived classes (ValueStateHandler & KeyStateHandler). As you have seen previously, there are a couple of value & key related states (WAITING_KEY, READING_KEY, WAITING_VALUE, READING_VALUE and more...). These classes handle those states. StateHandler, the base class, is just a class that contains the common parts of both. They are: `escape_character`, `enclosing_character`, `get() interface` and `createElement`.

Do you have a suggestion here?

vdimir replied:
Ok, let's just add some comments about `get` method and probably give a bit more specific name what we get (not state, but some data).

arthurpassos replied:
That does make sense, looking from another perspective `get` gives the idea of `get state`. I'll rename it to `getElement`, but can go for a different name if you have a better suggestion.

vdimir commented on src/Functions/keyvaluepair/src/KeyValuePairExtractor.h:
```diff
@@ -0,0 +1,32 @@
+ * Control characters (key_value_pair_separator, item_delimiter, escape_character and enclosing_character) are customizable
+ * */
+template <typename R = std::unordered_map<std::string, std::string>>
+struct KeyValuePairExtractor
```
Seems that `LazyEscapingKeyValuePairExtractor` is only inheritor, can't we not use this interface at all and expose just function `std::unordered_map<std::string_view, std::string_view> exctractKeyValuePairs(const std::string & file, KeyValueSettings settings)` ?

vdimir commented on src/Functions/keyvaluepair/src/KeyValuePairExtractor.h:
```diff
@@ -0,0 +1,32 @@
+
+    virtual ~KeyValuePairExtractor() = default;
+
+    virtual Response extract(const std::string & file) = 0;
```
Is file just data, not filepath?

arthurpassos replied:
I'll rename it to data

vdimir commented on src/Functions/extractKeyValuePairs.cpp:
```diff
@@ -0,0 +1,271 @@
+        return ParsedArguments{data_column, {}, {}, {}, {}, value_special_characters_allow_list};
+    }
+
+    auto escape_character = arguments[1].column->getDataAt(0).toView().front();
```
We may also check that all setting arguments (1-5) are constants.

arthurpassos replied:
Can you give me an example of that? Not sure what that means

vdimir replied:
I mean in sql arguments except first are typically constants `extractKeyValuePairs(some_column, '\\', ':', ',', '"')` and 
we don't support different chars for different rows (it would be confusing), so we can behave as `throwIf` do.

https://github.com/ClickHouse/ClickHouse/blob/c27f0d28a17e2aa71884a17c859a6c280597d8a6/src/Functions/throwIf.cpp#L69

```sql
select throwIf(0, 'message');
-- Ok

select throwIf(0, materialize('message'));
-- Code: 44. DB::Exception: Argument at index 1 for function throwIf must be constant. (ILLEGAL_COLUMN)

select throwIf(0, 'message' || toString(number)) from numbers(10);
-- Code: 44. DB::Exception: Argument at index 1 for function throwIf must be constant. (ILLEGAL_COLUMN)
```

arthurpassos commented on src/Functions/CMakeLists.txt:
```diff
@@ -103,6 +103,9 @@ endif()
 list (APPEND PRIVATE_LIBS clickhouse_functions_jsonpath)
 
+add_subdirectory(keyvaluepair)
+list (APPEND PRIVATE_LIBS clickhouse_functions_extractkeyvaluepairs)
```
I am not sure, I just followed the above JSONPath example. Do you want me to change it to `list (APPEND OBJECT_LIBS clickhouse_functions_extractkeyvaluepairs)` ?

arthurpassos commented on tests/queries/0_stateless/02499_extract_key_value_pairs_multiple_input.sql:
```diff
@@ -0,0 +1,29 @@
+-- { echoOn }
+----- Enclosed elements tests -----
+
+-- Expected output: {'name':'Neymar','team':'psg','age':'30','favorite_movie':'','height':'1.75'}
+SELECT extractKeyValuePairs('"name": "Neymar", "age": 30, team: "psg", "favorite_movie": "", height: 1.75', '\\', ':', ',', '"', '.');
```
That indeed looks better, didn't know it was possible!

arthurpassos replied:
Ok, that requires every single quote to be escaped. Unless there is another way, I think it's better to keep it the way it is

vdimir replied:
Indeed, `{}` syntax for maps is not supported, only `map('name', 'Neymar', 'team', 'psg', ...)`

arthurpassos commented on src/Functions/extractKeyValuePairs.cpp:
```diff
@@ -0,0 +1,230 @@
+
+    using ReplaceString = ReplaceStringImpl<ReplaceStringTraits::Replace::All>;
+
+    ReplaceString::vector(raw_keys->getChars(), raw_keys->getOffsets(), "\\", "", escaped_keys->getChars(), escaped_keys->getOffsets());
```
I am bit confused with docs. Should I have both (.md and this one)?

vdimir replied:
We aim to move docs into source code, so one passed to `registerFunction` is the source of truth. Having a duplicate in `md` for now is ok.

arthurpassos committed:
improve lib cmakefiles, add state docs and minor changes

arthurpassos commented on src/Functions/keyvaluepair/src/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+add_library(clickhouse_functions_extractkeyvaluepairs KeyValuePairExtractor.h impl/state/KeyStateHandler.cpp impl/state/KeyStateHandler.h impl/state/State.h impl/state/ValueStateHandler.cpp impl/state/ValueStateHandler.h impl/state/StateHandler.cpp impl/state/StateHandler.h impl/SimpleKeyValuePairEscapingProcessor.h impl/SimpleKeyValuePairEscapingProcessor.cpp KeyValuePairExtractorBuilder.h KeyVa ...
```
I have changed it to the below, could you check if that's what you want?

```
include("${ClickHouse_SOURCE_DIR}/cmake/dbms_glob_sources.cmake")
add_headers_and_sources(clickhouse_functions_extractkeyvaluepairs .)
add_headers_and_sources(clickhouse_functions_extractkeyvaluepairs impl)
add_headers_and_sources(clickhouse_functions_extractkeyvaluepairs impl/state)

add_library(clickhouse_functions_extractkeyvaluepairs ${clickhouse_functions_extractkeyvaluepairs_sources} ${clickhouse_functions_extractkeyvaluepairs_headers})
```

arthurpassos committed:
style

arthurpassos commented on src/Functions/keyvaluepair/src/KeyValuePairExtractorBuilder.h:
```diff
@@ -0,0 +1,78 @@
+
+        if (!escaping_processor)
+        {
+            throw std::runtime_error{"Escaping processor must be set, cannot build KeyValuePairExtractor without one"};
```
How do you feel about the builder throwing `std::runtime_error` and `DB::ExtractKeyValuePairs::getExtractor` catch it and do `throw DB::Exception(ErrorCodes::LOGICAL_ERROR, ex.what());`.

As of now, extraction library has no dependencies on CH. This would be the first one.

arthurpassos committed:
address comments

arthurpassos committed:
add control characters validation

arthurpassos committed:
sort output to make reference files more reliable

arthurpassos committed:
move to object libs

arthurpassos committed:
fix escape function

arthurpassos committed:
fix escape function

arthurpassos committed:
escp

arthurpassos commented:
@vdimir I know I still have to fix AST fuzzer, I did spend some time on it but couldn't find the issue. Will probably work on it again next week. Aside from that, is there anything else that you would like to be changed?

vdimir commented on tests/queries/0_stateless/02499_extract_key_value_pairs_multiple_input.sql:
```diff
@@ -0,0 +1,103 @@
+
+-- Expected output: {'name':'Neymar','team':'psg','age':'30','favorite_movie':'','height':'1.75'}
+WITH
+    extractKeyValuePairs('"name": "Neymar", "age": 30, team: "psg", "favorite_movie": "", height: 1.75', '\\', ':', ',', '"', '.') AS s_map,
```
Let's also add tests with some arguments omitted to check the defaults.

vdimir reviewed (commented):
Mostly alright, let's investigate failure

arthurpassos committed:
Remove state from Extractor. It's usually a bad idea that bites us in the future, it did.

arthurpassos committed:
alphanumeric as key start

arthurpassos commented:
The fuzz crash has been solved, but I'll make some more changes to improve the performance. Will let you know when it's good for review again

arthurpassos committed:
tmp

arthurpassos committed:
mess, i guess all bugs are fixed and performance a bit better

arthurpassos committed:
partially fix tests

arthurpassos committed:
well wlel, context switching

arthurpassos committed:
performance improvements and optional escaping

arthurpassos committed:
didnt understand

arthurpassos committed:
fix empty special value character allowlist

arthurpassos committed:
remove logical error extern

arthurpassos committed:
fix test

arthurpassos committed:
Merge branch 'master' into parsey_key_value_function

arthurpassos committed:
remove stale test / not reproducible anymore since I lost the python script

arthurpassos committed:
touching up

filimonov commented on src/Functions/keyvaluepair/src/impl/state/NoEscapingValueStateHandler.h:
```diff
@@ -0,0 +1,115 @@
+
+        value = {};
+
+        while (pos < file.size())
+        {
+            const auto current_character = file[pos++];
+
+            if (current_character == item_delimiter || !isValidCharacter(current_character))
+            {
+                value = createElement(file, start_index, pos - 1);
+                return {pos, State::FLUSH_PAIR};
+            }
+        }
+
+        // TODO: do I really need the below logic?
+        // this allows empty values at the end
+        value = createElement(file, start_index, pos);
+        return {pos, State::FLUSH_PAIR};
```
avoid unnessesary repetions, do smth like

```suggestion
        while (pos < file.size())
        {
            const auto current_character = file[pos++];

            if (current_character == item_delimiter || !isValidCharacter(current_character))
                  continue;
         }

        value = createElement(file, start_index, pos);
        return {pos, State::FLUSH_PAIR};
```

filimonov commented on src/Functions/keyvaluepair/tests/gtest_key_value_pair_extractor.cpp:
```diff
@@ -0,0 +1,167 @@
+//#include <Functions/keyvaluepair/src/StaleKeyValuePairExtractorBuilder.h>
+//#include <gtest/gtest.h>
+//
+//namespace DB
```
why everything is commented out? Maybe remove the test at all?

filimonov commented on src/Functions/keyvaluepair/src/impl/state/State.h:
```diff
@@ -0,0 +1,37 @@
+    // In this state, both key and value have already been collected and should be flushed. Might jump to WAITING_KEY or END.
+    FLUSH_PAIR,
+    END
+};
```
diagram of states (mermaid)

```mermaid
flowchart LR
    WAITING_KEY
    READING_KEY
    READING_ENCLOSED_KEY
    READING_KV_DELIMITER
    WAITING_VALUE
    READING_VALUE
    READING_ENCLOSED_VALUE
    READING_EMPTY_VALUE
    FLUSH_PAIR
    END

    WAITING_KEY -->|isalnum| READING_KEY
    WAITING_KEY -->|enclosing_character| READING_ENCLOSED_KEY
    WAITING_KEY -->|escape_character| READING_ENCLOSED_KEY
    WAITING_KEY -->|eof| END

    READING_KEY -->|key_value_delimiter| WAITING_VALUE
    READING_KEY -->|!isalnum| WAITING_KEY
    READING_KEY -->|current_character != '_'| WAITING_KEY
    READING_KEY -->|eof| END

    READING_ENCLOSED_KEY --> |enclosing_character| READING_KV_DELIMITER

    READING_KV_DELIMITER --> |key_value_delimiter| WAITING_VALUE
    READING_KV_DELIMITER --> |!key_value_delimiter| WAITING_KEY
    READING_KV_DELIMITER -->|eof| END

    WAITING_VALUE --> |enclosing_character| READING_ENCLOSED_VALUE 
    WAITING_VALUE --> |item_delimiter| READING_EMPTY_VALUE
    WAITING_VALUE --> |isValidCharacter| READING_VALUE
    WAITING_VALUE -->|eof| READING_EMPTY_VALUE


    READING_VALUE --> |item_delimiter| FLUSH_PAIR
    READING_VALUE --> |!isValidCharacter| FLUSH_PAIR
    READING_VALUE --> |eof| FLUSH_PAIR

    READING_EMPTY_VALUE --> FLUSH_PAIR

    FLUSH_PAIR --> WAITING_KEY 

    READING_ENCLOSED_VALUE --> |enclosing_character| FLUSH_PAIR
    READING_ENCLOSED_VALUE --> |eof| END
```

filimonov commented on docs/en/sql-reference/functions/string-functions.md:
```diff
@@ -1196,3 +1196,53 @@ Result:
+* key_value_pair = key kv_separator value
+* key = <quoted_string> |  asciichar asciialphanumeric*
+* kv_separator = ':'
+* value = <quoted_string> | asciialphanum*
```
I don't see <quoted_string> explained in that grammar

I don't see value_special_characters_allow_list used in that grammar 

I don't see escape_character used in that grammar

filimonov commented on tests/queries/0_stateless/02499_extract_key_value_pairs_multiple_input.sql:
```diff
@@ -0,0 +1,102 @@
+                ),
+            'Map(String,String)'
+        ) AS x
+SELECT
```
Very poor test coverage. 
Only few happypaths.

arthurpassos committed:
small refactoring

arthurpassos committed:
temp

arthurpassos committed:
Merge branch 'ClickHouse:master' into extract_key_value_pairs_find_delimiter_approach

arthurpassos committed:
Merge branch 'ClickHouse:master' into parsey_key_value_function

arthurpassos committed:
tmp

ilejn commented on src/Functions/keyvaluepair/src/impl/state/InlineEscapingKeyStateHandler.cpp:
```diff
@@ -0,0 +1,117 @@
+
+    while (pos < file.size())
+    {
+        const auto current_character = file[pos++];
```
Why not use string_view::find to eliminate this loop?
It is quite possible that find is more effective because of SSE or other optimizations.

arthurpassos replied:
This code is outdated, I am working on a new implementation that'll fundamentally change how the input is processed.

It'll most likely make use of `find_first_symbols`, which has SSE4 in it. Will also consider `string_view::find`

arthurpassos committed:
seems functional

arthurpassos committed:
Merge branch 'extract_key_value_pairs_find_delimiter_approach' into parsey_key_value_function

arthurpassos committed:
refactor to find_first logic instead of character validation

arthurpassos commented:
Published a VERY draft update, code is still VERY messy and will be improved.

Next changes from the top of my head:

1. Improve test coverage
2. Implement NoEscaping strategy
3. Implement multiple quoting characters support (as of now, even though it's accepting a list of characters, it does not work as expected if multiple are provided)
4. Input parameter validation
5. Think of a better solution for needle construction
6. Possibly improve code re-usage

arthurpassos committed:
add a few more tests

arthurpassos committed:
remove unwanted files

arthurpassos committed:
add first noescaping impl, add tests

arthurpassos committed:
remove support for multiple quoting characters

arthurpassos committed:
input parameter validation, still draft

arthurpassos committed:
cross check parameters

arthurpassos committed:
add some more input validation

arthurpassos committed:
minor changes

arthurpassos commented:
From the top of my head, pending things are:

1. More tests
2. Optimizations
3. Code clean-up

arthurpassos committed:
remove all usages of the word enclosed and use quoted instead

arthurpassos committed:
more tests

arthurpassos committed:
update reference

arthurpassos committed:
remove stale reference

arthurpassos committed:
validate pair delimiters does not contain escape character

arthurpassos committed:
remove reading_empty state

arthurpassos committed:
remove needles size check inside find_first_symbols

arthurpassos committed:
move runtime find_first_symbols to find_symbols.h

arthurpassos committed:
Merge branch 'master' into parsey_key_value_function

arthurpassos committed:
add a few tests for runtime find_symbols, more are needed

arthurpassos committed:
use standard parseComplexEscapedSequence instead of custom implementation. Drops support for 1 digit hex and octal digits

arthurpassos committed:
remove isEscapeCharacter method

arthurpassos committed:
style

Enmk committed:
Updated implementation of find_first_symbols for run-time needle

Implemented compile-time and run-time dispatching between SSE4.2 and SSE2 implementations
Added find_first_symbols_sse2
Added tests

Enmk committed:
Fixed implementation of find_first_symbols_sse42 run-time needle

Added some tests

Enmk committed:
Less noisy tests

Enmk committed:
Almost works

Enmk committed:
It works

Enmk committed:
Updated test case

Invalid escape sequence is removed from key or value, while corresponding key or value are not discarded

Enmk committed:
Minor optimization of copying data into key

Enmk committed:
Fused KeyStateHandler and ValueState togther

Removed some unneeded files and moved code around.

Enmk committed:
Minor adjustments

Enmk committed:
Changes discussed with Arthur

Enmk committed:
Fixed unit-tests

Enmk committed:
Implemented writing to ColumnString directly via StringWriter

arthurpassos committed:
Merge pull request #2 from Enmk/parsey_key_value_function-refactoring

Parsey key value function refactoring

arthurpassos committed:
fix string writer ub, escape sequence parsing failure and make state machine fully functional (remove extra if statements)

arthurpassos committed:
fixed another stringwriter bug, where multi-row input would break because of prev_commit_offset=0

arthurpassos committed:
validated output against main test case, performance is not the same as before, but still good

arthurpassos committed:
unit tests passing

arthurpassos committed:
refactor state handlers so writer comes from them

arthurpassos committed:
Use different string writers strategies for escaping and non escaping

arthurpassos committed:
extract configuration comparisons into meaningful methods

arthurpassos committed:
style

arthurpassos committed:
fix undefined mm_is_in_prepare

arthurpassos committed:
fix sse42 not defined build

arthurpassos committed:
remove big input test as it's outdated

arthurpassos committed:
remove debug find functions and replace SearchSymbols assertion with an exception, check with upstream preferred way

arthurpassos committed:
fix unit tests

arthurpassos committed:
fix clangtidy build

arthurpassos committed:
update docs

arthurpassos committed:
more docs

arthurpassos commented on base/base/find_symbols.h:
```diff
@@ -34,10 +35,52 @@
+#if defined(__SSE4_2__)
+        if (str.size() > BUFFER_SIZE)
+        {
+            throw std::runtime_error("SearchSymbols can contain at most " + std::to_string(BUFFER_SIZE) + " symbols and " + std::to_string(str.size()) + " was provided\n");
```
I believe `DB::Exception` should not be thrown here since it's in the `base` directory. Should it be a `std::exception` or a custom exception?

arthurpassos replied:
Are you sure about that? I do not see any `DB::Exception` being thrown in this directory, only custom defined ones. I also see the below comment on `base/throwError.h`. 

`/// DB::Exception is not defined yet. It'd better to throw Poco::Exception but we do not want to include any big header here, even <string>.`

vdimir replied:
Is this code expected to be unreachable? If yes, `std::runtime_error` is fine

arthurpassos replied:
SSE42 expects needle / search symbol buffer to contain 16 bytes. Hence, the exception. I would say it's reachable.

arthurpassos committed:
remove stale test file

arthurpassos committed:
more docs / comments

arthurpassos committed:
more docs

arthurpassos committed:
Use boolean column as escaping support argument instead of ColumnString

arthurpassos committed:
update docs to reflect new boolean argument

arthurpassos committed:
=default for extractkvp constructor

vdimir commented on base/base/find_symbols.h:
```diff
@@ -211,6 +362,11 @@ inline char * find_first_symbols(char * begin, char * end)
     return const_cast<char *>(detail::find_first_symbols_dispatch<true, detail::ReturnMode::End, symbols...>(begin, end));
 }
 
+inline const char * find_first_symbols(std::string_view haystack, const SearchSymbols & symbols)
```
Consider using methods from `std::`, please, if possible.
Are these functions introduced only for performance reasons?

arthurpassos replied:
Yes, for performance reasons. According my "silly" benchmarks, SSE4 presents better performance when keys and values are bigger.

Plus, these functions already exist in a compile-time version. Do you think we can keep it?

Enmk replied:
According [to a benchmark](https://quick-bench.com/#eyJ0ZXh0IjoiI3ByYWdtYSBjbGFuZyBhdHRyaWJ1dGUgcHVzaCAoX19hdHRyaWJ1dGVfXygodGFyZ2V0KFwic3NlLHNzZTIsc3NlNC4yXCIpKSksIGFwcGx5X3RvPWZ1bmN0aW9uKVxuXG4jaW5jbHVkZSA8Y3N0ZGludD5cbiNpbmNsdWRlIDxzdHJpbmc+XG4jaW5jbHVkZSA8c3RyaW5nX3ZpZXc+XG4jaW5jbHVkZSA8c3RyaW5nLmg+XG5cbiNpbmNsdWRlIDxhbGdvcml0aG0+XG4jaW5jbHVkZSA8ZXhlY3V0aW9uPlxuI2luY2x1ZGUgPHR5cGVfdHJhaXRzPlxuXG5cbi8vI2lmIGRlZmluZWQoX19TU0UyX18pXG4gICAgI2luY2x1ZGUgPGVtbWludHJpbi5oPlxuLy8jZW5kaWZcbi8vI2lmIGRlZmluZWQoX19TU0U0XzJfXylcbiAgICAjaW5jbHVkZSA8bm1taW50cmluLmg+XG4vLyNlbmRpZlxuXG5cblxuc3RydWN0IFNlYXJjaFN5bWJvbHNcbntcbiAgICBzdGF0aWMgY29uc3RleHByIGF1dG8gQlVGRkVSX1NJWkUgPSAxNjtcblxuICAgIFNlYXJjaFN5bWJvbHMoKSA9IGRlZmF1bHQ7XG5cbiAgICBTZWFyY2hTeW1ib2xzKHN0ZDo6c3RyaW5nX3ZpZXcgaW4pXG4gICAgICAgIDogc3RyKGluLnN1YnN0cigwLCBzdGQ6Om1pbjxzaXplX3Q+KEJVRkZFUl9TSVpFLCBpbi5zaXplKCkpKSlcbiAgICB7XG4vLyNpZiBkZWZpbmVkKF9fU1NFNF8yX18pXG4vLyAgICAgICAgaWYgKHN0ci5zaXplKCkgPiBCVUZGRVJfU0laRSlcbi8vICAgICAgICB7XG4vLyAgICAgICAgICAgIHRocm93IHN0ZDo6cnVudGltZV9lcnJvcihcIlNlYXJjaFN5bWJvbHMgY2FuIGNvbnRhaW4gYXQgbW9zdCBcIiArIHN0ZDo6dG9fc3RyaW5nKEJVRkZFUl9TSVpFKSArIFwiIHN5bWJvbHMgYW5kIFwiICsgc3RkOjp0b19zdHJpbmcoc3RyLnNpemUoKSkgKyBcIiB3YXMgcHJvdmlkZWRcXG5cIik7XG4vLyAgICAgICAgfVxuXG4gICAgICAgIGNoYXIgdG1wX3NhZmV0eV9idWZmZXJbQlVGRkVSX1NJWkVdID0gezB9O1xuXG4gICAgICAgIG1lbWNweSh0bXBfc2FmZXR5X2J1ZmZlciwgc3RyLmRhdGEoKSwgc3RyLnNpemUoKSk7XG5cbiAgICAgICAgc2ltZF92ZWN0b3IgPSBfbW1fbG9hZHVfc2kxMjgocmVpbnRlcnByZXRfY2FzdDxjb25zdCBfX20xMjhpICo+KHRtcF9zYWZldHlfYnVmZmVyKSk7XG4vLyNlbmRpZlxuICAgIH1cblxuLy8jaWYgZGVmaW5lZChfX1NTRTRfMl9fKVxuICAgIF9fbTEyOGkgc2ltZF92ZWN0b3I7XG4vLyNlbmRpZlxuICAgIHN0ZDo6c3RyaW5nX3ZpZXcgc3RyO1xufTtcblxubmFtZXNwYWNlIGRldGFpbFxue1xudGVtcGxhdGUgPGNoYXIgLi4uY2hhcnM+IGNvbnN0ZXhwciBib29sIGlzX2luKGNoYXIgeCkgeyByZXR1cm4gKCh4ID09IGNoYXJzKSB8fCAuLi4pOyB9XG5cbnN0YXRpYyBib29sIGlzX2luKGNoYXIgYywgY29uc3QgY2hhciAqIHN5bWJvbHMsIHNpemVfdCBudW1fY2hhcnMpXG57XG4gICAgZm9yIChzaXplX3QgaSA9IDB1OyBpIDwgbnVtX2NoYXJzOyArK2kpXG4gICAge1xuICAgICAgICBpZiAoYyA9PSBzeW1ib2xzW2ldKVxuICAgICAgICB7XG4gICAgICAgICAgICByZXR1cm4gdHJ1ZTtcbiAgICAgICAgfVxuICAgIH1cblxuICAgIHJldHVybiBmYWxzZTtcbn1cblxuaW5saW5lIF9fbTEyOGkgbW1faXNfaW4oX19tMTI4aSBieXRlcywgY29uc3QgY2hhciAqIHN5bWJvbHMsIHNpemVfdCBudW1fY2hhcnMpXG57XG4gICAgX19tMTI4aSBhY2N1bXVsYXRvciA9IF9tbV9zZXR6ZXJvX3NpMTI4KCk7XG4gICAgZm9yIChzaXplX3QgaSA9IDA7IGkgPCBudW1fY2hhcnM7ICsraSlcbiAgICB7XG4gICAgICAgIF9fbTEyOGkgZXEgPSBfbW1fY21wZXFfZXBpOChieXRlcywgX21tX3NldDFfZXBpOChzeW1ib2xzW2ldKSk7XG4gICAgICAgIGFjY3VtdWxhdG9yID0gX21tX29yX3NpMTI4KGFjY3VtdWxhdG9yLCBlcSk7XG4gICAgfVxuXG4gICAgcmV0dXJuIGFjY3VtdWxhdG9yO1xufVxuXG5pbmxpbmUgc3RkOjphcnJheTxfX20xMjhpLCAxNnU+IG1tX2lzX2luX3ByZXBhcmUoY29uc3QgY2hhciAqIHN5bWJvbHMsIHNpemVfdCBudW1fY2hhcnMpXG57XG4gICAgc3RkOjphcnJheTxfX20xMjhpLCAxNnU+IHJlc3VsdCB7fTtcblxuICAgIGZvciAoc2l6ZV90IGkgPSAwOyBpIDwgbnVtX2NoYXJzOyArK2kpXG4gICAge1xuICAgICAgICByZXN1bHRbaV0gPSBfbW1fc2V0MV9lcGk4KHN5bWJvbHNbaV0pO1xuICAgIH1cblxuICAgIHJldHVybiByZXN1bHQ7XG59XG5cbmlubGluZSBfX20xMjhpIG1tX2lzX2luX2V4ZWN1dGUoX19tMTI4aSBieXRlcywgY29uc3Qgc3RkOjphcnJheTxfX20xMjhpLCAxNnU+ICYgbmVlZGxlcylcbntcbiAgICBfX20xMjhpIGFjY3VtdWxhdG9yID0gX21tX3NldHplcm9fc2kxMjgoKTtcblxuICAgIGZvciAoY29uc3QgYXV0byAmIG5lZWRsZSA6IG5lZWRsZXMpXG4gICAge1xuICAgICAgICBfX20xMjhpIGVxID0gX21tX2NtcGVxX2VwaTgoYnl0ZXMsIG5lZWRsZSk7XG4gICAgICAgIGFjY3VtdWxhdG9yID0gX21tX29yX3NpMTI4KGFjY3VtdWxhdG9yLCBlcSk7XG4gICAgfVxuXG4gICAgcmV0dXJuIGFjY3VtdWxhdG9yO1xufVxuXG50ZW1wbGF0ZSA8Ym9vbCBwb3NpdGl2ZT5cbmNvbnN0ZXhwciBib29sIG1heWJlX25lZ2F0ZShib29sIHgpIHsgcmV0dXJuIHggPT0gcG9zaXRpdmU7IH1cblxudGVtcGxhdGUgPGJvb2wgcG9zaXRpdmU+XG5jb25zdGV4cHIgdWludDE2X3QgbWF5YmVfbmVnYXRlKHVpbnQxNl90IHgpXG57XG4gICAgaWYgY29uc3RleHByIChwb3NpdGl2ZSlcbiAgICAgICAgcmV0dXJuIHg7XG4gICAgZWxzZVxuICAgICAgICByZXR1cm4gfng7XG59XG5cbmVudW0gY2xhc3MgUmV0dXJuTW9kZVxue1xuICAgIEVuZCxcbiAgICBOdWxscHRyLFxufTtcblxuXG50ZW1wbGF0ZSA8Ym9vbCBwb3NpdGl2ZSwgUmV0dXJuTW9kZSByZXR1cm5fbW9kZT5cbmlubGluZSBjb25zdCBjaGFyICogZmluZF9maXJzdF9zeW1ib2xzX3NzZTIoY29uc3QgY2hhciAqIGNvbnN0IGJlZ2luLCBjb25zdCBjaGFyICogY29uc3QgZW5kLCBjb25zdCBjaGFyICogc3ltYm9scywgc2l6ZV90IG51bV9jaGFycylcbntcbiAgICBjb25zdCBjaGFyICogcG9zID0gYmVnaW47XG5cbiNpZiBkZWZpbmVkKF9fU1NFMl9fKVxuICAgIGNvbnN0IGF1dG8gbmVlZGxlcyA9IG1tX2lzX2luX3ByZXBhcmUoc3ltYm9scywgbnVtX2NoYXJzKTtcbiAgICBmb3IgKDsgcG9zICsgMTUgPCBlbmQ7IHBvcyArPSAxNilcbiAgICB7XG4gICAgICAgIF9fbTEyOGkgYnl0ZXMgPSBfbW1fbG9hZHVfc2kxMjgocmVpbnRlcnByZXRfY2FzdDxjb25zdCBfX20xMjhpICo+KHBvcykpO1xuXG4gICAgICAgIF9fbTEyOGkgZXEgPSBtbV9pc19pbl9leGVjdXRlKGJ5dGVzLCBuZWVkbGVzKTtcblxuICAgICAgICB1aW50MTZfdCBiaXRfbWFzayA9IG1heWJlX25lZ2F0ZTxwb3NpdGl2ZT4odWludDE2X3QoX21tX21vdmVtYXNrX2VwaTgoZXEpKSk7XG4gICAgICAgIGlmIChiaXRfbWFzaylcbiAgICAgICAgICAgIHJldHVybiBwb3MgKyBfX2J1aWx0aW5fY3R6KGJpdF9tYXNrKTtcbiAgICB9XG4jZW5kaWZcblxuICAgIGZvciAoOyBwb3MgPCBlbmQ7ICsrcG9zKVxuICAgICAgICBpZiAobWF5YmVfbmVnYXRlPHBvc2l0aXZlPihpc19pbigqcG9zLCBzeW1ib2xzLCBudW1fY2hhcnMpKSlcbiAgICAgICAgICAgIHJldHVybiBwb3M7XG5cbiAgICByZXR1cm4gcmV0dXJuX21vZGUgPT0gUmV0dXJuTW9kZTo6RW5kID8gZW5kIDogbnVsbHB0cjtcbn1cblxuXG50ZW1wbGF0ZSA8Ym9vbCBwb3NpdGl2ZSwgUmV0dXJuTW9kZSByZXR1cm5fbW9kZT5cbmlubGluZSBjb25zdCBjaGFyICogZmluZF9maXJzdF9zeW1ib2xzX3NzZTQyKGNvbnN0IGNoYXIgKiBjb25zdCBiZWdpbiwgY29uc3QgY2hhciAqIGNvbnN0IGVuZCwgY29uc3QgU2VhcmNoU3ltYm9scyAmIHN5bWJvbHMpXG57XG4gICAgY29uc3QgY2hhciAqIHBvcyA9IGJlZ2luO1xuXG4gICAgY29uc3QgYXV0byBudW1fY2hhcnMgPSBzeW1ib2xzLnN0ci5zaXplKCk7XG5cbi8vI2lmIGRlZmluZWQoX19TU0U0XzJfXylcbiAgICBjb25zdGV4cHIgaW50IG1vZGUgPSBfU0lERF9VQllURV9PUFMgfCBfU0lERF9DTVBfRVFVQUxfQU5ZIHwgX1NJRERfTEVBU1RfU0lHTklGSUNBTlQ7XG5cbiAgICBjb25zdCBfX20xMjhpIHNldCA9IHN5bWJvbHMuc2ltZF92ZWN0b3I7XG5cbiAgICBmb3IgKDsgcG9zICsgMTUgPCBlbmQ7IHBvcyArPSAxNilcbiAgICB7XG4gICAgICAgIF9fbTEyOGkgYnl0ZXMgPSBfbW1fbG9hZHVfc2kxMjgocmVpbnRlcnByZXRfY2FzdDxjb25zdCBfX20xMjhpICo+KHBvcykpO1xuXG4gICAgICAgIGlmIGNvbnN0ZXhwciAocG9zaXRpdmUpXG4gICAgICAgIHtcbiAgICAgICAgICAgIGlmIChfbW1fY21wZXN0cmMoc2V0LCBudW1fY2hhcnMsIGJ5dGVzLCAxNiwgbW9kZSkpXG4gICAgICAgICAgICAgICAgcmV0dXJuIHBvcyArIF9tbV9jbXBlc3RyaShzZXQsIG51bV9jaGFycywgYnl0ZXMsIDE2LCBtb2RlKTtcbiAgICAgICAgfVxuICAgICAgICBlbHNlXG4gICAgICAgIHtcbiAgICAgICAgICAgIGlmIChfbW1fY21wZXN0cmMoc2V0LCBudW1fY2hhcnMsIGJ5dGVzLCAxNiwgbW9kZSB8IF9TSUREX05FR0FUSVZFX1BPTEFSSVRZKSlcbiAgICAgICAgICAgICAgICByZXR1cm4gcG9zICsgX21tX2NtcGVzdHJpKHNldCwgbnVtX2NoYXJzLCBieXRlcywgMTYsIG1vZGUgfCBfU0lERF9ORUdBVElWRV9QT0xBUklUWSk7XG4gICAgICAgIH1cbiAgICB9XG4vLyNlbmRpZlxuXG4gICAgZm9yICg7IHBvcyA8IGVuZDsgKytwb3MpXG4gICAgICAgIGlmIChtYXliZV9uZWdhdGU8cG9zaXRpdmU+KGlzX2luKCpwb3MsIHN5bWJvbHMuc3RyLmRhdGEoKSwgbnVtX2NoYXJzKSkpXG4gICAgICAgICAgICByZXR1cm4gcG9zO1xuXG4gICAgcmV0dXJuIHJldHVybl9tb2RlID09IFJldHVybk1vZGU6OkVuZCA/IGVuZCA6IG51bGxwdHI7XG59XG5cbi8vLyBOT1RFIE5vIFNTRSA0LjIgaW1wbGVtZW50YXRpb24gZm9yIGZpbmRfbGFzdF9zeW1ib2xzX29yX251bGwuIE5vdCB3b3J0aCB0byBkby5cblxuXG50ZW1wbGF0ZSA8Ym9vbCBwb3NpdGl2ZSwgUmV0dXJuTW9kZSByZXR1cm5fbW9kZT5cbmlubGluZSBjb25zdCBjaGFyICogZmluZF9maXJzdF9zeW1ib2xzX2Rpc3BhdGNoKGNvbnN0IHN0ZDo6c3RyaW5nX3ZpZXcgaGF5c3RhY2ssIGNvbnN0IFNlYXJjaFN5bWJvbHMgJiBzeW1ib2xzKVxue1xuLy8jaWYgZGVmaW5lZChfX1NTRTRfMl9fKVxuICAgIGlmIChzeW1ib2xzLnN0ci5zaXplKCkgPj0gNSlcbiAgICAgICAgcmV0dXJuIGZpbmRfZmlyc3Rfc3ltYm9sc19zc2U0Mjxwb3NpdGl2ZSwgcmV0dXJuX21vZGU+KGhheXN0YWNrLmJlZ2luKCksIGhheXN0YWNrLmVuZCgpLCBzeW1ib2xzKTtcbiAgICBlbHNlXG4vLyNlbmRpZlxuICAgICAgICByZXR1cm4gZmluZF9maXJzdF9zeW1ib2xzX3NzZTI8cG9zaXRpdmUsIHJldHVybl9tb2RlPihoYXlzdGFjay5iZWdpbigpLCBoYXlzdGFjay5lbmQoKSwgc3ltYm9scy5zdHIuZGF0YSgpLCBzeW1ib2xzLnN0ci5zaXplKCkpO1xufVxuXG59XG5cblxuXG5pbmxpbmUgY29uc3QgY2hhciAqIGZpbmRfZmlyc3Rfc3ltYm9scyhzdGQ6OnN0cmluZ192aWV3IGhheXN0YWNrLCBjb25zdCBTZWFyY2hTeW1ib2xzICYgc3ltYm9scylcbntcbiAgICByZXR1cm4gZGV0YWlsOjpmaW5kX2ZpcnN0X3N5bWJvbHNfZGlzcGF0Y2g8dHJ1ZSwgZGV0YWlsOjpSZXR1cm5Nb2RlOjpFbmQ+KGhheXN0YWNrLCBzeW1ib2xzKTtcbn1cblxuXG5cbmlubGluZSBjb25zdCBjaGFyICogZmluZF9maXJzdF9ub3Rfc3ltYm9scyhzdGQ6OnN0cmluZ192aWV3IGhheXN0YWNrLCBjb25zdCBTZWFyY2hTeW1ib2xzICYgc3ltYm9scylcbntcbiAgICByZXR1cm4gZGV0YWlsOjpmaW5kX2ZpcnN0X3N5bWJvbHNfZGlzcGF0Y2g8ZmFsc2UsIGRldGFpbDo6UmV0dXJuTW9kZTo6RW5kPihoYXlzdGFjaywgc3ltYm9scyk7XG59XG5cbmlubGluZSBjb25zdCBjaGFyICogZmluZF9maXJzdF9zeW1ib2xzX29yX251bGwoc3RkOjpzdHJpbmdfdmlldyBoYXlzdGFjaywgY29uc3QgU2VhcmNoU3ltYm9scyAmIHN5bWJvbHMpXG57XG4gICAgcmV0dXJuIGRldGFpbDo6ZmluZF9maXJzdF9zeW1ib2xzX2Rpc3BhdGNoPHRydWUsIGRldGFpbDo6UmV0dXJuTW9kZTo6TnVsbHB0cj4oaGF5c3RhY2ssIHN5bWJvbHMpO1xufVxuXG5cbmlubGluZSBjb25zdCBjaGFyICogZmluZF9maXJzdF9ub3Rfc3ltYm9sc19vcl9udWxsKHN0ZDo6c3RyaW5nX3ZpZXcgaGF5c3RhY2ssIGNvbnN0IFNlYXJjaFN5bWJvbHMgJiBzeW1ib2xzKVxue1xuICAgIHJldHVybiBkZXRhaWw6OmZpbmRfZmlyc3Rfc3ltYm9sc19kaXNwYXRjaDxmYWxzZSwgZGV0YWlsOjpSZXR1cm5Nb2RlOjpOdWxscHRyPihoYXlzdGFjaywgc3ltYm9scyk7XG59XG5cblxudXNpbmcgbmFtZXNwYWNlIHN0ZDo6bGl0ZXJhbHM7XG5cbnN0YXRpYyBjb25zdCBhdXRvIGhheXN0YWNrID0gUlwidGV4dChcbjEpIFRoZSBleGVjdXRpb24gcG9saWN5IHR5cGUgdXNlZCBhcyBhIHVuaXF1ZSB0eXBlIHRvIGRpc2FtYmlndWF0ZSBwYXJhbGxlbCBhbGdvcml0aG0gb3ZlcmxvYWRpbmcgYW5kIHJlcXVpcmUgdGhhdCBhIHBhcmFsbGVsIGFsZ29yaXRobSdzIGV4ZWN1dGlvbiBtYXkgbm90IGJlIHBhcmFsbGVsaXplZC4gVGhlIGludm9jYXRpb25zIG9mIGVsZW1lbnQgYWNjZXNzIGZ1bmN0aW9ucyBpbiBwYXJhbGxlbCBhbGdvcml0aG1zIGludm9rZWQgd2l0aCB0aGlzIHBvbGljeSAodXN1YWxseSBzcGVjaWZpZWQgYXMgc3RkOjpleGVjdXRpb246OnNlcSkgYXJlIGluZGV0ZXJtaW5hdGVseSBzZXF1ZW5jZWQgaW4gdGhlIGNhbGxpbmcgdGhyZWFkLlxuMikgVGhlIGV4ZWN1dGlvbiBwb2xpY3kgdHlwZSB1c2VkIGFzIGEgdW5pcXVlIHR5cGUgdG8gZGlzYW1iaWd1YXRlIHBhcmFsbGVsIGFsZ29yaXRobSBvdmVybG9hZGluZyBhbmQgaW5kaWNhdGUgdGhhdCBhIHBhcmFsbGVsIGFsZ29yaXRobSdzIGV4ZWN1dGlvbiBtYXkgYmUgcGFyYWxsZWxpemVkLiBUaGUgaW52b2NhdGlvbnMgb2YgZWxlbWVudCBhY2Nlc3MgZnVuY3Rpb25zIGluIHBhcmFsbGVsIGFsZ29yaXRobXMgaW52b2tlZCB3aXRoIHRoaXMgcG9saWN5ICh1c3VhbGx5IHNwZWNpZmllZCBhcyBzdGQ6OmV4ZWN1dGlvbjo6cGFyKSBhcmUgcGVybWl0dGVkIHRvIGV4ZWN1dGUgaW4gZWl0aGVyIHRoZSBpbnZva2luZyB0aHJlYWQgb3IgaW4gYSB0aHJlYWQgaW1wbGljaXRseSBjcmVhdGVkIGJ5IHRoZSBsaWJyYXJ5IHRvIHN1cHBvcnQgcGFyYWxsZWwgYWxnb3JpdGhtIGV4ZWN1dGlvbi4gQW55IHN1Y2ggaW52b2NhdGlvbnMgZXhlY3V0aW5nIGluIHRoZSBzYW1lIHRocmVhZCBhcmUgaW5kZXRlcm1pbmF0ZWx5IHNlcXVlbmNlZCB3aXRoIHJlc3BlY3QgdG8gZWFjaCBvdGhlci5cbjMpIFRoZSBleGVjdXRpb24gcG9saWN5IHR5cGUgdXNlZCBhcyBhIHVuaXF1ZSB0eXBlIHRvIGRpc2FtYmlndWF0ZSBwYXJhbGxlbCBhbGdvcml0aG0gb3ZlcmxvYWRpbmcgYW5kIGluZGljYXRlIHRoYXQgYSBwYXJhbGxlbCBhbGdvcml0aG0ncyBleGVjdXRpb24gbWF5IGJlIHBhcmFsbGVsaXplZCwgdmVjdG9yaXplZCwgb3IgbWlncmF0ZWQgYWNyb3NzIHRocmVhZHMgKHN1Y2ggYXMgYnkgYSBwYXJlbnQtc3RlYWxpbmcgc2NoZWR1bGVyKS4gVGhlIGludm9jYXRpb25zIG9mIGVsZW1lbnQgYWNjZXNzIGZ1bmN0aW9ucyBpbiBwYXJhbGxlbCBhbGdvcml0aG1zIGludm9rZWQgd2l0aCB0aGlzIHBvbGljeSBhcmUgcGVybWl0dGVkIHRvIGV4ZWN1dGUgaW4gYW4gdW5vcmRlcmVkIGZhc2hpb24gaW4gdW5zcGVjaWZpZWQgdGhyZWFkcywgYW5kIHVuc2VxdWVuY2VkIHdpdGggcmVzcGVjdCB0byBvbmUgYW5vdGhlciB3aXRoaW4gZWFjaCB0aHJlYWQuXG40KSBUaGUgZXhlY3V0aW9uIHBvbGljeSB0eXBlIHVzZWQgYXMgYSB1bmlxdWUgdHlwZSB0byBkaXNhbWJpZ3VhdGUgcGFyYWxsZWwgYWxnb3JpdGhtIG92ZXJsb2FkaW5nIGFuZCBpbmRpY2F0ZSB0aGF0IGEgcGFyYWxsZWwgYWxnb3JpdGhtJ3MgZXhlY3V0aW9uIG1heSBiZSB2ZWN0b3JpemVkLCBlLmcuLCBleGVjdXRlZCBvbiBhIHNpbmdsZSB0aHJlYWQgdXNpbmcgaW5zdHJ1Y3Rpb25zIHRoYXQgb3BlcmF0ZSBvbiBtdWx0aXBsZSBkYXRhIGl0ZW1zLlxufiFAIyQlXiYqKCkvXG4pdGV4dFwic3Y7XG5cbi8vIEhlcmUgd2UgbWFrZSBzdXJlIG5lZWRsZXMgYXJlIGF0IGxlYXN0IDE2IGJ5dGVzIGxvbmcsIHdlJ2xsIGNyZWF0ZSBzaG9ydGVyIG5lZWRsZXMgZnJvbSB0aGF0IGxhdGVyXG5zdGF0aWMgY29uc3Qgc3RkOjpzdHJpbmdfdmlldyB0ZXN0X25lZWRsZXNbXSA9IHtcbiAgICBcImBfXFxcImBfXFxcImBfXFxcImBfXFxcImBfXFxcImBfXFxcImBfXFxcImBfXFxcImBfXFxcImBfXFxcImBfXFxcImBfXFxcImBfXFxcImBfXFxcImBfXFxcIlwic3YsIC8vIGRvZXNuJ3QgbWF0Y2hcbiAgICBoYXlzdGFjay5zdWJzdHIoaGF5c3RhY2suc2l6ZSgpIC0gNjQsIGhheXN0YWNrLnNpemUoKSksXG4gICAgaGF5c3RhY2suc3Vic3RyKGhheXN0YWNrLnNpemUoKSAtIDMyLCBoYXlzdGFjay5zaXplKCkpLFxuICAgIGhheXN0YWNrLnN1YnN0cihoYXlzdGFjay5zaXplKCkgLSAxNiwgaGF5c3RhY2suc2l6ZSgpKSxcbn07XG5cbmF1dG8gbWF4X3NpemVfbmVlZGxlcyhjb25zdCBhdXRvICYgaW5fbmVlZGxlcywgc2l6ZV90IG1heF9zaXplKSB7XG4gICAgc3RkOjp2ZWN0b3I8c3RkOjpkZWNheV90PGRlY2x0eXBlKGluX25lZWRsZXNbMF0pPj4gcmVzdWx0O1xuICAgIHJlc3VsdC5yZXNpemUoc3RkOjpzaXplKGluX25lZWRsZXMpKTtcblxuICAgIGZvciAoY29uc3QgYXV0byAmIG4gOiBpbl9uZWVkbGVzKSB7XG4gICAgICAgIHJlc3VsdC5wdXNoX2JhY2sobi5zdWJzdHIoc3RkOjptaW48c2l6ZV90PihuLnNpemUoKSwgbWF4X3NpemUpKSk7XG4gICAgfVxuXG4gICAgcmV0dXJuIHJlc3VsdDtcbn1cblxuc3RhdGljIHZvaWQgQk1fZmZzX1NUTChiZW5jaG1hcms6OlN0YXRlJiBzdGF0ZSkge1xuXG4gICAgY29uc3QgYXV0byBuZWVkbGVzID0gbWF4X3NpemVfbmVlZGxlcyh0ZXN0X25lZWRsZXMsIHN0YXRlLnJhbmdlKDApKTtcblxuICAgIGZvciAoYXV0byBfIDogc3RhdGUpIHtcbiAgICAgICAgZm9yIChjb25zdCBhdXRvICYgbmVlZGxlIDogbmVlZGxlcykge1xuICAgICAgICAgICAgYmVuY2htYXJrOjpEb05vdE9wdGltaXplKFxuICAgICAgICAgICAgICAgIHN0ZDo6ZmluZF9maXJzdF9vZihzdGQ6OmV4ZWN1dGlvbjo6dW5zZXEsIGhheXN0YWNrLmJlZ2luKCksIGhheXN0YWNrLmVuZCgpLCBuZWVkbGUuYmVnaW4oKSwgbmVlZGxlLmVuZCgpKVxuICAgICAgICAgICAgKTtcbiAgICAgICAgfVxuICAgIH1cbn1cblxuXG5zdGF0aWMgdm9pZCBCTV9mZnNfU1NFKGJlbmNobWFyazo6U3RhdGUmIHN0YXRlKSB7XG4gIC8vIFBlcmZvcm0gc2V0dXAgaGVyZVxuICAgIGNvbnN0IGF1dG8gbmVlZGxlcyA9IG1heF9zaXplX25lZWRsZXModGVzdF9uZWVkbGVzLCBzdGF0ZS5yYW5nZSgwKSk7XG5cbiAgICBzdGQ6OnZlY3RvcjxTZWFyY2hTeW1ib2xzPiBwcmVwYXJlZF9uZWVkZWxlcztcbiAgICBwcmVwYXJlZF9uZWVkZWxlcy5yZXNlcnZlKHN0ZDo6c2l6ZShuZWVkbGVzKSk7XG4gICAgZm9yIChjb25zdCBhdXRvICYgbmVlZGxlIDogbmVlZGxlcylcbiAgICAgICAgcHJlcGFyZWRfbmVlZGVsZXMuZW1wbGFjZV9iYWNrKG5lZWRsZSk7XG5cbiAgICBmb3IgKGF1dG8gXyA6IHN0YXRlKSB7XG4gICAgICAgIGZvciAoY29uc3QgYXV0byAmIG5lZWRsZSA6IHByZXBhcmVkX25lZWRlbGVzKVxuICAgICAgICAgICAgYmVuY2htYXJrOjpEb05vdE9wdGltaXplKFxuICAgICAgICAgICAgICAgIGZpbmRfZmlyc3Rfc3ltYm9sc19vcl9udWxsKGhheXN0YWNrLCBuZWVkbGUpXG4gICAgICAgICAgICApO1xuICAgIH1cbn1cblxuQkVOQ0hNQVJLKEJNX2Zmc19TVEwpLT5BcmcoMiktPkFyZyg0KS0+QXJnKDgpLT5BcmcoMTYpO1xuQkVOQ0hNQVJLKEJNX2Zmc19TU0UpLT5BcmcoMiktPkFyZyg0KS0+QXJnKDgpLT5BcmcoMTYpO1xuXG5cbiNwcmFnbWEgY2xhbmcgYXR0cmlidXRlIHBvcCIsImNvbXBpbGVyIjoiY2xhbmctMTQuMCIsIm9wdGltIjoiMiJ9) `find_first_symbols_or_null` is significantly faster than corresponding `std::find_first_of(std::execution::unseq, ...`

![image](https://user-images.githubusercontent.com/1110183/230782111-451be529-24b1-4dc9-853b-fc2fd3f0d0d0.png)

Enmk commented on src/Functions/keyvaluepair/extractKeyValuePairs.cpp:
```diff
@@ -0,0 +1,213 @@
+            A key-value pair consists of a key followed by a `key_value_delimiter` and a value. Quoted keys and values are also supported. Key value pairs must be separated by pair delimiters.
+            Escaping support can be turned on and off.
+
+            Escape sequences supported: `\x`, `\N`, `\a`, `\b`, `\e`, `\f`, `\n`, `\r`, `\t`, `\v` and `\0`.
```
Does it really still supports octal numerals? **\`\0\`** or is it just an outdated comment?

arthurpassos replied:
Null-terminator character

Enmk reviewed (changes requested):
Mostly Ok, but IMO few changes required.

Bool arguments are not quite user-friendly, so I'd suggest having a separate function name instead (e.g. `extractKeyValueEscaped`). 
Rest of clickhouse uses suffixes to distinguish function variants, see 'OrNull' and 'orDefault':
https://clickhouse.com/docs/en/sql-reference/functions/type-conversion-functions#toint8163264128256orzero

arthurpassos committed:
minor renaming

arthurpassos committed:
Use two functions extractKeyValuePairs and extractKeyValuePairsWithEscaping instead of boolean parameter

arthurpassos commented:
> Mostly Ok, but IMO few changes required.
> 
> Bool arguments are not quite user-friendly, so I'd suggest having a separate function name instead (e.g. `extractKeyValueEscaped`). Rest of clickhouse uses suffixes to distinguish function variants, see 'OrNull' and 'orDefault': https://clickhouse.com/docs/en/sql-reference/functions/type-conversion-functions#toint8163264128256orzero

Added two functions: `extractKeyValuePairs` and `extractKeyValuePairsWithEscaping`

arthurpassos committed:
remove escaping_support parameter left overs

arthurpassos committed:
use std::array instead of std::vector

arthurpassos committed:
remove extra backslash in docs

arthurpassos committed:
Merge branch 'master' into parsey_key_value_function

arthurpassos committed:
add input arguments type validation

arthurpassos committed:
reject values starting with escape sequences if escaping support is turned on

arthurpassos committed:
define ILLEGAL_TYPE_OF_ARGUMENT in ArgumentExtractor

arthurpassos committed:
fix escap sequences unit tests, values are not allowed to start with escape sequences anymore

arthurpassos committed:
Merge branch 'master' into parsey_key_value_function

arthurpassos commented:
Hi @vdimir, I have now marked it as ready for review. First of all, I apologize for re-requesting your review. I have summarized the changes from the last version you reviewed to this one. Please let me know if you have any question, will answer them as quickly as I can!

arthurpassos committed:
Add validation for number of arguments provided and update doc examples

arthurpassos committed:
minor doc update

arthurpassos commented:
As requested offline by @vdimir, I'll be attaching the results of a benchmark comparing queries against a build using SSE42/2 string search and a build using `std::find_first_of`:

#### Environment

Builds were generated from this code by tweaking AABA:

```
#define AABA 0

inline auto find(std::string_view file, const SearchSymbols & needles)
{
#if AABA == 1
    return std::find_first_of(file.begin(), file.end(), needles.str.begin(), needles.str.end());
#else
    return find_first_symbols_or_null(file, needles);
#endif
}
```

Table used:

```
arthur :) show create table bif

SHOW CREATE TABLE bif

Query id: 13a5dedd-be82-42a3-b9a0-e03b95555cd6

┌─statement────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ CREATE TABLE default.bif
(
    raw String,
    kv Map(String, String) ALIAS extractKeyValuePairs(raw, '=', ' ,|-%')
)
ENGINE = MergeTree
ORDER BY tuple()
SETTINGS index_granularity = 8192 │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.001 sec. 

arthur :)
```

File used to fill in the table: [big_input_file.txt](https://github.com/ClickHouse/ClickHouse/files/11276440/big_input_file.txt)


On port 9000, the SSE42/2 build is running. On 9001, std::find_first_of.

#### Results

```
arthur@arthur:~/sse4_test$ clickhouse-benchmark --query "select kv from bif" -i 100 --host=localhost --port=9000 --host=localhost --port=9001
Loaded 1 queries.

Queries executed: 100.

localhost:9000, queries 56, QPS: 277.317, RPS: 277.317, MiB/s: 608.092, result RPS: 277.317, result MiB/s: 165.644.
localhost:9001, queries 44, QPS: 146.915, RPS: 146.915, MiB/s: 322.150, result RPS: 146.915, result MiB/s: 87.754.

0.000%		0.003 sec.	0.006 sec.	
10.000%		0.003 sec.	0.006 sec.	
20.000%		0.003 sec.	0.007 sec.	
30.000%		0.003 sec.	0.007 sec.	
40.000%		0.003 sec.	0.007 sec.	
50.000%		0.004 sec.	0.007 sec.	
60.000%		0.004 sec.	0.007 sec.	
70.000%		0.004 sec.	0.007 sec.	
80.000%		0.004 sec.	0.007 sec.	
90.000%		0.004 sec.	0.007 sec.	
95.000%		0.004 sec.	0.008 sec.	
99.000%		0.005 sec.	0.009 sec.	
99.900%		0.005 sec.	0.009 sec.	
99.990%		0.005 sec.	0.009 sec.	

Difference at 99.5% confidence: mean difference is 0.00320068, but confidence interval is 0.00032690
```

arthurpassos committed:
minor doc update

arthurpassos committed:
Merge branch 'master' into parsey_key_value_function

vdimir commented on tests/queries/0_stateless/02499_extract_key_value_pairs_multiple_input.sql:
```diff
@@ -0,0 +1,428 @@
+-- invalid escape sequence, should be discarded
+-- expected output: {'valid_key':'valid_value'}
+WITH
+    extractKeyValuePairsWithEscaping('valid_key:valid_value key:invalid_val\\', ':', ' ', '"') AS s_map,
```
Why don't we treat `\\` at the end as garbage and not ignore it and parse `key:invalid_val` ?

arthurpassos replied:
Done

vdimir commented on src/IO/ReadHelpers.h:
```diff
@@ -1578,4 +1578,7 @@ void readQuotedField(String & s, ReadBuffer & buf);
 
 void readTSVField(String & s, ReadBuffer & buf);
+
+bool parseComplexEscapeSequence(String & s, ReadBuffer & buf);
```
Please add a comment, what exactly is cosidered as `EscapeSequence` what is `Complex` one.

arthurpassos replied:
Ok. This method already exists, it just does not have a public API. I'll copy its implementation comment if it's alright:

```
/** Parse the escape sequence, which can be simple (one character after backslash) or more complex (multiple characters).
  * It is assumed that the cursor is located on the `\` symbol
  */
```

vdimir commented on src/Functions/keyvaluepair/impl/StateHandler.h:
```diff
@@ -0,0 +1,51 @@
+        WAITING_VALUE,
+        // Tries to read a value. Jumps to FLUSH_PAIR.
+        READING_VALUE,
+        // Tries to read a quoted/ quoted value. Might jump to FLUSH_PAIR or END.
```
Typo?

`Tries to read a quoted/ quoted`

vdimir commented on src/Functions/keyvaluepair/extractKeyValuePairs.cpp:
````diff
@@ -0,0 +1,226 @@
+
+            **Syntax**
+            ``` sql
+            extractKeyValuePairs(data, [key_value_delimiter], [pair_delimiter], [quoting_character])
````
We should also decide how to unify it with another PR with `str_to_map`. Maybe `str_to_map` (or `mapFromString` can be just an alias to exctractKeyValuePair). Then the question is about arguments.

E.g., in https://docs.databricks.com/sql/language-manual/functions/str_to_map.html there're `pairDelim` in 2nd position and `keyValueDelim` in 3rd. Another similar function is https://github.com/GoogleCloudPlatform/bigquery-utils/tree/master/udfs/community#cw_map_parsem-string-pd-string-kvd-string

Should we also do like this? Do you have any opinion on the order of arguments?

arthurpassos replied:
In [DataDog's parseKeyValue](https://docs.datadoghq.com/logs/log_configuration/parsing/?tab=matchers#key-value-or-logfmt) function, key value delimiter is the 2nd. That seems to be a very minor thing for me and would require re-ordering all parameters in code and tests.

arthurpassos replied:
But if you think it's the best order, I can do it. No problem :)

arthurpassos commented on src/IO/ReadHelpers.cpp:
```diff
@@ -336,7 +336,23 @@ static ReturnType parseComplexEscapeSequence(Vector & s, ReadBuffer & buf)
+            {
+                readPODBinary(hex_code, buf);
+            }
+            catch (const Exception &)
```
Hm, that is an interesting question. I initially thought it was an error when I saw that `parseComplexEscapeSequence<String, bool>` could throw even though `ReturnType` is set not to throw. I don't quite remember if there were side effects of not catching all exceptions, tho.

Basically, it's a matter of design / expectations about the function. Do you have a strong opinion on this?

vdimir replied:
I believe there're exceptions that we don't want to catch. ReadBuffer can be reader from remote source and it can be network error. Or error on memory allocation.

arthurpassos committed:
address comments and update docs

arthurpassos committed:
docs

arthurpassos committed:
fix comment

arthurpassos committed:
docs

arthurpassos committed:
fix ut
